File: app/src/main/java/top/asdb/codexremote/agent/RemoteAgentClient.kt
package top.asdb.codexremote.agent

import java.io.InputStream
import java.io.OutputStream
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.serialization.json.JsonObject
import top.asdb.codexremote.data.AgentConnectionTestResult
import top.asdb.codexremote.data.AgentGlobalSettings
import top.asdb.codexremote.data.AgentModel
import top.asdb.codexremote.data.AgentThread
import top.asdb.codexremote.data.AgentCapabilities
import top.asdb.codexremote.data.AgentKind
import top.asdb.codexremote.data.ApiModelOption
import top.asdb.codexremote.data.ApprovalMode
import top.asdb.codexremote.data.ApprovalPrompt
import top.asdb.codexremote.data.CustomModelDefinition
import top.asdb.codexremote.data.PendingAttachment
import top.asdb.codexremote.data.RemoteDirectoryListing
import top.asdb.codexremote.data.RemoteFileListing
import top.asdb.codexremote.data.RemoteFileTransferMode
import top.asdb.codexremote.data.ServerMetrics
import top.asdb.codexremote.data.ServerProfile
import top.asdb.codexremote.data.ThreadGoal
import top.asdb.codexremote.data.ThreadGoalStatus
import top.asdb.codexremote.data.TimelineEntry
import top.asdb.codexremote.data.TokenUsage
import top.asdb.codexremote.ssh.RemoteInstallProgress

/** Protocol-neutral notification consumed by the shared conversation reducer. */
data class AgentNotification(
    val generation: Long,
    val method: String,
    val params: JsonObject,
    val sequence: Long = 0,
)

data class AgentApproval(val generation: Long, val prompt: ApprovalPrompt)

data class AgentConnectionEvent(val generation: Long, val message: String)

/** Result of probing one agent runtime on an SSH host. */
data class AgentRuntimeInspection(
    val os: String,
    val architecture: String,
    val home: String,
    val detectedVersion: String? = null,
    val compatibleCommand: String? = null,
    val installationProblem: String? = null,
)

class UnsupportedAgentCapabilityException(
    agent: AgentKind,
    capability: String,
) : UnsupportedOperationException("${agent.label} 暂不支持$capability")

/**
 * Contract between shared app state/UI and a concrete remote coding agent.
 *
 * Adapters own protocol mapping and runtime startup. The ViewModel depends only on this contract,
 * so a future agent can be added without another profile/session/connection implementation.
 */
interface RemoteAgentClient {
    val kind: AgentKind
    val capabilities: AgentCapabilities

    val notifications: SharedFlow<AgentNotification>
    val approvals: SharedFlow<AgentApproval>
    val diagnostics: SharedFlow<AgentConnectionEvent>
    val closed: SharedFlow<AgentConnectionEvent>

    suspend fun probeFingerprint(profile: ServerProfile): String
    suspend fun inspectRuntime(profile: ServerProfile): AgentRuntimeInspection
    suspend fun installRuntime(
        profile: ServerProfile,
        onProgress: (RemoteInstallProgress) -> Unit,
    )
    suspend fun uninstallRuntime(profile: ServerProfile)

    suspend fun connect(profile: ServerProfile): String
    suspend fun disconnect()
    fun close()
    fun isConnected(): Boolean
    fun currentGeneration(): Long?
    fun isGenerationActive(generation: Long): Boolean
    fun isClosedGenerationCurrent(generation: Long): Boolean

    suspend fun listThreads(search: String = "", archived: Boolean = false): List<AgentThread>
    suspend fun listModels(): List<AgentModel>
    suspend fun openThread(threadId: String, approvalMode: ApprovalMode): AgentSession
    suspend fun listThreadTurnsPage(
        threadId: String,
        cursor: String,
        subAgentCreatedAt: Long? = null,
    ): AgentThreadPage
    suspend fun readThread(threadId: String): Pair<AgentThread, List<TimelineEntry>>
    fun cachedThread(threadId: String): AgentThreadCacheSnapshot?
    fun cachedThreadStale(threadId: String): AgentThreadCacheSnapshot?
    fun cachedContextUsage(threadId: String): TokenUsage?
    fun cacheThread(
        thread: AgentThread,
        timeline: List<TimelineEntry>,
        nextTurnsCursor: String? = null,
        tokenUsage: TokenUsage? = null,
    )

    suspend fun startThread(
        profile: ServerProfile,
        model: String?,
        approvalMode: ApprovalMode,
    ): Pair<AgentThread, List<TimelineEntry>>
    suspend fun startTurn(
        threadId: String,
        text: String,
        attachments: List<PendingAttachment>,
        model: String?,
        effort: String?,
        approvalMode: ApprovalMode,
        cwd: String?,
    ): String

    /**
     * Makes a locally-added model available to the remote Agent before it is selected.
     * Adapters that do not need provider-side model registration can keep the no-op default.
     */
    suspend fun ensureCustomModel(
        profile: ServerProfile,
        definition: CustomModelDefinition,
    ): Unit = Unit

    /**
     * Reconciles locally-managed models with the remote Agent configuration. Removed IDs are
     * explicit so an adapter never has to infer ownership from the provider's complete catalog.
     */
    suspend fun syncCustomModels(
        profile: ServerProfile,
        definitions: List<CustomModelDefinition>,
        removedModelIds: List<String>,
    ) {
        definitions.forEach { ensureCustomModel(profile, it) }
    }

    suspend fun steerTurn(
        threadId: String,
        turnId: String,
        text: String,
        attachments: List<PendingAttachment>,
    ): Unit = unsupported("回合引导")
    suspend fun interruptTurn(threadId: String, turnId: String)

    suspend fun compactThread(threadId: String): Unit = unsupported("上下文压缩")
    suspend fun getThreadGoal(threadId: String): ThreadGoal? = null
    suspend fun setThreadGoal(
        threadId: String,
        objective: String? = null,
        status: ThreadGoalStatus? = null,
        tokenBudget: Long? = null,
    ): ThreadGoal = unsupported("会话目标")
    suspend fun clearThreadGoal(threadId: String): Unit = unsupported("会话目标")
    suspend fun archiveThread(threadId: String)
    suspend fun rollbackThread(
        threadId: String,
        approvalMode: ApprovalMode,
        turns: Int = 1,
    ): AgentSession = unsupported("回滚")
    suspend fun setThreadName(threadId: String, name: String)
    suspend fun startReview(threadId: String): Unit = unsupported("代码审查")

    suspend fun listDirectories(path: String?): RemoteDirectoryListing
    suspend fun listFiles(path: String?): RemoteFileListing
    suspend fun readServerMetrics(profile: ServerProfile): ServerMetrics
    suspend fun upload(name: String, bytes: ByteArray): String
    suspend fun uploadFile(directory: String, name: String, input: InputStream)
    suspend fun downloadFile(path: String, output: OutputStream)
    suspend fun renameFile(path: String, newName: String)
    suspend fun deleteFiles(paths: List<String>)
    suspend fun transferFiles(
        paths: List<String>,
        destinationDirectory: String,
        mode: RemoteFileTransferMode,
    )
    suspend fun downloadImage(path: String): ByteArray

    suspend fun readGlobalSettings(profile: ServerProfile): AgentGlobalSettings =
        unsupported("全局设置")
    suspend fun writeGlobalSettings(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
        defaultModel: String,
        defaultReasoningEffort: String,
        preserveCurrentProvider: Boolean,
    ): Unit = unsupported("全局设置")
    suspend fun testGlobalSettings(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
        testModel: String,
    ): AgentConnectionTestResult = unsupported("全局设置")
    suspend fun fetchApiModels(
        profile: ServerProfile,
        baseUrl: String,
        apiKey: String,
        proxyUrl: String,
    ): List<ApiModelOption> = unsupported("模型 API")
    suspend fun answerApproval(
        prompt: ApprovalPrompt,
        accept: Boolean,
        answers: Map<String, String> = emptyMap(),
    ): Unit = unsupported("操作审批")

    private fun unsupported(capability: String): Nothing =
        throw UnsupportedAgentCapabilityException(kind, capability)
}

File: flutter_app/lib/src/agent/remote_agent_client.dart
import '../domain/models.dart';
import '../ssh/ssh_server_client.dart';
import 'codex_protocol.dart';
import 'remote_bootstrap.dart';

/// A page returned by an agent's thread-list endpoint.
class AgentThreadPage {
  const AgentThreadPage({
    required this.threads,
    this.nextCursor,
    this.previousCursor,
  });

  final List<AgentThread> threads;
  final String? nextCursor;
  final String? previousCursor;
}

class AgentSession {
  const AgentSession({
    required this.thread,
    required this.timeline,
    this.nextTurnsCursor,
    this.tokenUsage,
    this.responseSequence = 0,
    this.activeTurnStartedAtMillis,
    this.turnIds = const <String>[],
    this.itemsView = 'full',
  });

  final AgentThread thread;
  final List<TimelineEntry> timeline;
  final String? nextTurnsCursor;
  final TokenUsage? tokenUsage;
  final int responseSequence;
  final int? activeTurnStartedAtMillis;
  final List<String> turnIds;
  final String itemsView;
}

class AgentTurnsPage {
  const AgentTurnsPage({
    required this.timeline,
    this.nextCursor,
    this.turnIds = const <String>[],
    this.itemsView = 'full',
  });

  final List<TimelineEntry> timeline;
  final String? nextCursor;
  final List<String> turnIds;
  final String itemsView;
}

/// Optional capability used to bind resume-time buffering to the concrete
/// app-server generation without source-breaking lightweight test adapters.
abstract interface class RemoteAgentGenerationClient {
  int? get currentGeneration;
}

/// Events that are safe for the controller to consume without knowing the
/// wire format of a particular agent adapter.
sealed class RemoteAgentEvent {
  const RemoteAgentEvent();
}

class RemoteAgentNotification extends RemoteAgentEvent {
  const RemoteAgentNotification(this.message);

  final CodexRpcNotification message;
}

class RemoteAgentServerRequest extends RemoteAgentEvent {
  const RemoteAgentServerRequest(this.message);

  final CodexServerRequest message;
}

class RemoteAgentDiagnostic extends RemoteAgentEvent {
  const RemoteAgentDiagnostic(this.message, {this.isStderr = false});

  final String message;
  final bool isStderr;
}

class RemoteAgentConnectionLost extends RemoteAgentEvent {
  const RemoteAgentConnectionLost(this.message);

  final String message;
}

/// Adapter contract shared by Codex and future OpenCode implementations.
///
/// A client is bound to one `(profileId, AgentKind)` lane by
/// [AgentConnectionManager]. It must never share its app-server channel with
/// another lane.
abstract interface class RemoteAgentClient {
  AgentKind get kind;

  AgentCapabilities get capabilities;

  bool get isConnected;

  Stream<RemoteAgentEvent> get events;

  Future<void> connect(ServerProfile profile, RemoteServerClient host);

  Future<List<AgentModel>> listModels();

  Future<AgentThreadPage> listThreads({String? searchTerm});

  Future<AgentSession> resumeThread(
    String threadId, {
    ApprovalMode approvalMode = ApprovalMode.requestApproval,
  });

  Future<AgentTurnsPage> loadOlderTurns({
    required String threadId,
    required String cursor,
    int? subAgentCreatedAt,
  });

  Future<void> disconnect();

  void close();
}

/// Optional pagination capability for adapters whose thread/list endpoint
/// exposes a continuation cursor. Keeping this separate preserves the small
/// base contract for lightweight read-only adapters and test fakes.
abstract interface class RemoteAgentThreadPaginationClient {
  Future<AgentThreadPage> listThreadsPage({String? searchTerm, String? cursor});
}

/// Optional capability implemented by agents that can execute turns. Keeping
/// this separate lets read-only adapters and third-party test fakes continue
/// implementing [RemoteAgentClient] without a source-breaking change.
abstract interface class RemoteAgentTurnClient {
  Future<String> startTurn({
    required String threadId,
    required String text,
    List<PendingAttachment> attachments = const <PendingAttachment>[],
    String? model,
    String? effort,
    ApprovalMode approvalMode = ApprovalMode.requestApproval,
    SandboxChoice? sandbox,
    String? cwd,
  });

  /// Interrupts the active turn. Read-only adapters may leave this disabled.
  Future<void> interruptTurn({
    required String threadId,
    required String turnId,
  });
}

/// Optional capability for adding input to a turn that is already running.
abstract interface class RemoteAgentSteerClient {
  Future<void> steerTurn({
    required String threadId,
    required String turnId,
    required String text,
    List<PendingAttachment> attachments = const <PendingAttachment>[],
  });
}

abstract interface class RemoteAgentThreadCreateClient {
  Future<AgentSession> startThread({
    String? cwd,
    String? model,
    ApprovalMode approvalMode = ApprovalMode.requestApproval,
    SandboxChoice? sandbox,
  });
}

/// Optional capability for app-server initiated approvals and user input.
abstract interface class RemoteAgentApprovalClient {
  Future<void> answerApproval(
    ApprovalPrompt prompt, {
    required bool accept,
    Map<String, String> answers = const <String, String>{},
  });
}

/// Optional thread mutations shared by the Codex work-screen actions. Keeping
/// this capability separate means an adapter can still provide read-only
/// history without pretending to support destructive operations.
abstract interface class RemoteAgentThreadMutationClient {
  Future<void> compactThread(String threadId);

  Future<AgentSession> rollbackThread(
    String threadId, {
    ApprovalMode approvalMode = ApprovalMode.requestApproval,
    int turns = 1,
  });

  Future<void> archiveThread(String threadId);

  Future<void> setThreadName(String threadId, String name);

  Future<void> startReview(String threadId);

  Future<ThreadGoal?> getThreadGoal(String threadId);

  Future<ThreadGoal> setThreadGoal(
    String threadId, {
    String? objective,
    ThreadGoalStatus? status,
    int? tokenBudget,
  });

  Future<void> clearThreadGoal(String threadId);
}

/// Optional per-user configuration capability. Implementations must edit the
/// Agent's global user files, never a repository-local configuration.
abstract interface class RemoteAgentGlobalSettingsClient {
  Future<AgentGlobalSettings> readGlobalSettings(ServerProfile profile);

  Future<void> writeGlobalSettings(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
    required String defaultModel,
    required String defaultReasoningEffort,
    required bool preserveCurrentProvider,
  });

  Future<AgentConnectionTestResult> testGlobalSettings(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
    required String testModel,
  });
}

/// Optional capability for listing models exposed by the configured HTTP API.
/// The request runs through the connected server so its network and proxy
/// settings match the Agent runtime.
abstract interface class RemoteAgentApiModelClient {
  Future<List<ApiModelOption>> fetchApiModels(
    ServerProfile profile, {
    required String baseUrl,
    required String apiKey,
    required String proxyUrl,
  });
}

/// Optional lifecycle for adapters that must register app-defined models in
/// the remote provider before those model IDs can be used by a turn.
abstract interface class RemoteAgentCustomModelClient {
  Future<void> ensureCustomModel(
    ServerProfile profile,
    CustomModelDefinition definition,
  );

  Future<void> syncCustomModels(
    ServerProfile profile, {
    required List<CustomModelDefinition> definitions,
    required List<String> removedModelIds,
  });
}

/// Optional capability for agents whose pinned runtime can be provisioned in
/// the current SSH user's home directory.
abstract interface class RemoteAgentRuntimeClient {
  Future<AgentRuntimeInspection> inspectRuntime(
    ServerProfile profile,
    RemoteServerClient host,
  );

  Future<void> installRuntime(
    ServerProfile profile,
    RemoteServerClient host, {
    required void Function(RemoteInstallProgress progress) onProgress,
  });

  Future<void> uninstallRuntime(ServerProfile profile, RemoteServerClient host);
}

/// A deliberately small adapter for agents that have not been migrated yet.
class UnsupportedRemoteAgentClient implements RemoteAgentClient {
  UnsupportedRemoteAgentClient(this.kind);

  @override
  final AgentKind kind;

  @override
  AgentCapabilities get capabilities => AgentCapabilities.none;

  @override
  bool get isConnected => false;

  @override
  Stream<RemoteAgentEvent> get events => const Stream<RemoteAgentEvent>.empty();

  @override
  Future<void> connect(ServerProfile profile, RemoteServerClient host) async {
    throw StateError('${kind.label} 适配器尚未启用');
  }

  @override
  Future<List<AgentModel>> listModels() async => const <AgentModel>[];

  @override
  Future<AgentThreadPage> listThreads({String? searchTerm}) async =>
      const AgentThreadPage(threads: <AgentThread>[]);

  @override
  Future<AgentSession> resumeThread(
    String threadId, {
    ApprovalMode approvalMode = ApprovalMode.requestApproval,
  }) async => throw StateError('${kind.label} 适配器尚未启用');

  @override
  Future<AgentTurnsPage> loadOlderTurns({
    required String threadId,
    required String cursor,
    int? subAgentCreatedAt,
  }) async => throw StateError('${kind.label} 适配器尚未启用');

  @override
  Future<void> disconnect() async {}

  @override
  void close() {}
}
